presto-main/src/main/java/com/facebook/presto/cost/SimpleFilterProjectSemiJoinStatsRule.java

```diff
@@ -0,0 +1,162 @@
+    }
+
+    @Override
+    public Pattern getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Optional<PlanNodeStatsEstimate> calculate(PlanNode node, Lookup lookup, Session session, Map<Symbol, Type> types)
+    {
+        FilterNode filterNode = (FilterNode) node;
+        SemiJoinNode semiJoinNode;
+        if (filterNode.getSource() instanceof ProjectNode) {
+            ProjectNode projectNode = (ProjectNode) filterNode.getSource();
+            if (!projectNode.isIdentity()) {
+                return Optional.empty();
+            }
+            if (!(projectNode.getSource() instanceof SemiJoinNode)) {
+                return Optional.empty();
+            }
+            semiJoinNode = (SemiJoinNode) projectNode.getSource();
+        }
+        else {
+            if (!(filterNode.getSource() instanceof SemiJoinNode)) {
+                return Optional.empty();
+            }
+            semiJoinNode = (SemiJoinNode) filterNode.getSource();
+        }
+
+        return calculate(filterNode, semiJoinNode, lookup, session, types);
+    }
+
+    private Optional<PlanNodeStatsEstimate> calculate(FilterNode filterNode, SemiJoinNode semiJoinNode, Lookup lookup, Session session, Map<Symbol, Type> types)
+    {
+        PlanNodeStatsEstimate sourceStats = lookup.getStats(semiJoinNode.getSource(), session, types);
+        PlanNodeStatsEstimate filteringSourceStats = lookup.getStats(semiJoinNode.getFilteringSource(), session, types);
+        Symbol filteringSourceJoinSymbol = semiJoinNode.getFilteringSourceJoinSymbol();
+        Symbol sourceJoinSymbol = semiJoinNode.getSourceJoinSymbol();
+
+        Optional<SemiJoinOutputFilter> semiJoinOutputFilter = extractSemiJoinOutputFilter(filterNode.getPredicate(), semiJoinNode.getSemiJoinOutput());
+
+        if (!semiJoinOutputFilter.isPresent()) {
+            return Optional.empty();
+        }
+
+        PlanNodeStatsEstimate semiJoinStats;
+        if (semiJoinOutputFilter.get().isNegated()) {
+            semiJoinStats = semiJoinStatsCalculator.computeAntiJoin(sourceStats, filteringSourceStats, sourceJoinSymbol, filteringSourceJoinSymbol);
+        }
+        else {
+            semiJoinStats = semiJoinStatsCalculator.computeSemiJoin(sourceStats, filteringSourceStats, sourceJoinSymbol, filteringSourceJoinSymbol);
+        }
+
+        // apply remaining prediate
+        PlanNodeStatsEstimate intermediateFilterStats = filterStatsCalculator.filterStats(semiJoinStats, semiJoinOutputFilter.get().getRemainingPredicate(), session, types);
+
+        // filter out unneeded symbols
+        PlanNodeStatsEstimate.Builder resultStats = PlanNodeStatsEstimate.buildFrom(intermediateFilterStats);
+        intermediateFilterStats.getSymbolsWithKnownStatistics().stream()
+                .filter(s -> !filterNode.getOutputSymbols().contains(s))
+                .forEach(s -> resultStats.removeSymbolStatistics(s));
```

Comment:
There should be some `removeMany` or `retain(Collection symbols)` -- for performance reasons. AFAIU p-collections perform better with such bulk operations.